src/main/java/net/neoforged/neoforge/common/world/chunk/ForcedChunkManager.java

```diff
@@ -0,0 +1,369 @@
+import it.unimi.dsi.fastutil.longs.LongSet;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import javax.annotation.ParametersAreNonnullByDefault;
+import net.minecraft.core.BlockPos;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.ListTag;
+import net.minecraft.nbt.NbtUtils;
+import net.minecraft.nbt.Tag;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.TicketType;
+import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.ForcedChunksSavedData;
+import net.minecraft.world.level.saveddata.SavedData;
+import net.neoforged.fml.ModLoader;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.annotations.ApiStatus;
+
+@ParametersAreNonnullByDefault
+public class ForcedChunkManager {
+    private static final Logger LOGGER = LogManager.getLogger();
+    static final TicketType<TicketOwner<BlockPos>> BLOCK = TicketType.create("neoforge:block", Comparator.comparing(info -> info));
+    static final TicketType<TicketOwner<BlockPos>> BLOCK_TICKING = TicketType.create("neoforge:block_ticking", Comparator.comparing(info -> info));
+    static final TicketType<TicketOwner<UUID>> ENTITY = TicketType.create("neoforge:entity", Comparator.comparing(info -> info));
+    static final TicketType<TicketOwner<UUID>> ENTITY_TICKING = TicketType.create("neoforge:entity_ticking", Comparator.comparing(info -> info));
+
+    private static boolean initialised = false;
+    private static Map<ResourceLocation, TicketController> controllers = Map.of();
+
+    @ApiStatus.Internal
+    public static synchronized void init() {
+        if (initialised) {
+            throw new UnsupportedOperationException("Cannot init ticket controllers multiple times!");
+        }
+        initialised = true;
+
+        final Map<ResourceLocation, TicketController> controllers = new HashMap<>();
+        ModLoader.get().postEvent(new RegisterTicketControllersEvent(controller -> {
+            if (controllers.containsKey(controller.id())) {
+                throw new IllegalArgumentException("Attempted to register two controllers with the same ID " + controller.id());
+            }
+            controllers.put(controller.id(), controller);
+        }));
+        ForcedChunkManager.controllers = Map.copyOf(controllers);
+    }
+
+    /**
+     * Checks if a level has any forced chunks. Mainly used for seeing if a level should continue ticking with no players in it.
+     */
+    public static boolean hasForcedChunks(ServerLevel level) {
+        ForcedChunksSavedData data = level.getDataStorage().get(new SavedData.Factory<>(ForcedChunksSavedData::new, ForcedChunksSavedData::load), "chunks");
```

Comment:
shouldn't the factory instance be stored somewhere?

Reply:
Technically, but vanilla isn't doing it either. and it's a vanilla saved data.